### ISSUE TITLE
ao=wasapi:exclusive no audio on video with mono audio tracks

### ISSUE THREAD

**Issue by ebvjr** [mpv-player/mpv#1643]
mpv says couldn't find acceptable audio format but the audio is working propoerly on mpc-hc + reclock with no errors.

mpv -v log: https://gist.github.com/ebvjr/a5831cdd046713c66102

```javascript
General
Unique ID                      : 173903417723215382414484652572197232744 (0x82D494786B2A5971903508FAEAC3A068)
Complete name                  : C:\Downloads\[AnimeOut]_Ragnarok_-_01_[480p]_[ven].mkv
Format                         : Matroska
Format version                 : Version 4 / Version 2
File size                      : 55.1 MiB
Duration                       : 23mn 44s
Overall bit rate               : 324 Kbps
Encoded date                   : UTC 2015-01-12 04:16:04
Writing application            : mkvmerge v7.2.0 ('On Every Street') 32bit built on Sep 13 2014 15:42:11
Writing library                : libebml v1.3.0 + libmatroska v1.4.1
DURATION                       : 00:00:12.000000000
NUMBER_OF_FRAMES               : 2
NUMBER_OF_BYTES                : 94
_STATISTICS_WRITING_APP        : mkvmerge v7.2.0 ('On Every Street') 32bit built on Sep 13 2014 15:42:11
_STATISTICS_WRITING_DATE_UTC   : 2015-01-12 04:16:04
_STATISTICS_TAGS               : BPS DURATION NUMBER_OF_FRAMES NUMBER_OF_BYTES

Video
ID                             : 1
Format                         : AVC
Format/Info                    : Advanced Video Codec
Format profile                 : High@L3.0
Format settings, CABAC         : Yes
Format settings, ReFrames      : 6 frames
Codec ID                       : V_MPEG4/ISO/AVC
Duration                       : 23mn 44s
Nominal bit rate               : 225 Kbps
Width                          : 640 pixels
Height                         : 480 pixels
Display aspect ratio           : 4:3
Frame rate mode                : Constant
Frame rate                     : 23.810 fps
Color space                    : YUV
Chroma subsampling             : 4:2:0
Bit depth                      : 8 bits
Scan type                      : Progressive
Bits/(Pixel*Frame)             : 0.031
Writing library                : x264 core 142 r2479 dd79a61
Encoding settings              : cabac=1 / ref=6 / deblock=1:1:1 / analyse=0x3:0x113 / me=hex / subme=7 / psy=1 / psy_rd=0.40:0.00 / mixed_ref=1 / me_range=16 / chroma_me=1 / trellis=1 / 8x8dct=1 / cqm=0 / deadzone=21,11 / fast_pskip=1 / chroma_qp_offset=-2 / threads=12 / lookahead_threads=2 / sliced_threads=0 / nr=1000 / decimate=1 / interlaced=0 / bluray_compat=0 / constrained_intra=0 / bframes=5 / b_pyramid=2 / b_adapt=1 / b_bias=0 / direct=1 / weightb=1 / open_gop=0 / weightp=2 / keyint=240 / keyint_min=24 / scenecut=40 / intra_refresh=0 / rc_lookahead=40 / rc=2pass / mbtree=1 / bitrate=225 / ratetol=1.0 / qcomp=0.60 / qpmin=0 / qpmax=69 / qpstep=4 / cplxblur=20.0 / qblur=0.5 / ip_ratio=1.40 / aq=1:0.60
Default                        : Yes
Forced                         : No
Color primaries                : BT.601 NTSC
Transfer characteristics       : BT.709
Matrix coefficients            : BT.601

Audio #1
ID                             : 2
Format                         : AAC
Format/Info                    : Advanced Audio Codec
Format profile                 : LC
Codec ID                       : A_AAC
Duration                       : 23mn 44s
Channel(s)                     : 1 channel
Channel positions              : Front: C
Sampling rate                  : 48.0 KHz
Compression mode               : Lossy
Delay relative to video        : -22ms
Language                       : English
Default                        : Yes
Forced                         : No

Audio #2
ID                             : 3
Format                         : AAC
Format/Info                    : Advanced Audio Codec
Format profile                 : LC
Codec ID                       : A_AAC
Duration                       : 23mn 44s
Channel(s)                     : 1 channel
Channel positions              : Front: C
Sampling rate                  : 48.0 KHz
Compression mode               : Lossy
Delay relative to video        : -22ms
Language                       : Japanese
Default                        : No
Forced                         : No

Text
ID                             : 4
Format                         : ASS
Codec ID                       : S_TEXT/ASS
Codec ID/Info                  : Advanced Sub Station Alpha
Compression mode               : Lossless
Default                        : Yes
Forced                         : No
```
![wasapi_001](https://cloud.githubusercontent.com/assets/9087493/6458851/5e11e260-c1c3-11e4-848c-85ee7d7bd811.png)

**Comment by wm4** [comment]
I think this happens because the file is mono.

**Comment by ebvjr** [comment]
I though it was fixed though https://github.com/mpv-player/mpv/issues/1350. It's pretty weird when reclock wasapi:exclusive can read the audio and mpv can't.

**Comment by wm4** [comment]
> It's pretty weird when reclock wasapi:exclusive can read the audio and mpv can't.

Why is that weird? From what I remember, wasapi does not support a straight-forward way to handle this.

**Comment by Skibicki** [comment]
It works with --audio-channels=2.0

**Comment by kevmitch** [comment]
This is a known issue. Channel selection is the most problematic part of wasapi exclusive mode. 

Exclusive mode requires that we exhaustively test all possible formats until a supported one is found. You can see this happening with --msg-level=ao=debug

The algorithm starts with the number of channels of the source and if that doesn't work, reduces by one channel and tries again. I have independently discovered recently, that my sound device also won't accept raw mono. Therefore a mono source will fail and since there are no further channels to remove, the overall search will fail. 

Specifying --audio-channels overrides the initial channels tried and therefore setting the correct number here is a workaround.

As soon as I get a chance, I will modify it so that the search starts with the greater of 2 or the number of channels in the source.

**Comment by Kovensky** [comment]
You probably should try mono before trying stereo if the source is mono.

**Comment by kevmitch** [comment]
fixed in 41c10c3ec2c8706841e72100d715e064d78e0cdb
